optimum/executorchruntime/modeling_executorch.py

```diff
@@ -0,0 +1,158 @@
+"""ExecuTorchModelForXXX classes, allowing to run ExecuTorch Models with ExecuTorch Runtime using the same API as Transformers."""
+
+import logging
+import os
+import warnings
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Set, Tuple, Union
+
+import torch
+from executorch.extension.pybindings.portable_lib import _load_for_executorch
+from huggingface_hub import hf_hub_download
+from huggingface_hub.constants import HUGGINGFACE_HUB_CACHE
+from huggingface_hub.utils import EntryNotFoundError
+from transformers import (
+    AutoConfig,
+    AutoModel,
+    GenerationMixin,
+    AutoModelForCausalLM,
+    GenerationConfig,
+)
+from transformers.integrations.executorch import TorchExportableModuleWithStaticCache
+from transformers.modeling_outputs import (
+    BaseModelOutput,
+    CausalLMOutput,
+    CausalLMOutputWithPast,
+    ModelOutput,
+)
+
+from ..exporters import TasksManager
+from ..exporters.executorch import main_export
+from ..modeling_base import FROM_PRETRAINED_START_DOCSTRING, OptimizedModel
+
+if TYPE_CHECKING:
+    from transformers import PretrainedConfig
+
+
+logger = logging.getLogger(__name__)
+
+
+class ExecuTorchModelForCausalLM(OptimizedModel):
+    """
+    ExecuTorch model with a causal language modeling head for ExecuTorch Runtime inference.
+    """
+
+    auto_model_class = AutoModelForCausalLM
+
+    def __init__(
+        self,
+        model: "ExecuTorchModule",  # Similar to ort.InferenceSession for ORTModel
+        config: "PretrainedConfig",
+    ):
+        super().__init__(model, config)
+        # TODO: Remove once figure out how make `ExecuTorchModule` inherit from `PreTrainedModel`
+        self.et_model = model
+
+    def forward(self, input_ids: torch.Tensor, cache_position: torch.Tensor) -> torch.Tensor:
+        return self.et_model.forward(input_ids, cache_position)[0]
```

Comment:
@michaelbenayoun I had trouble to figure out how the runtime is integrated. It seems like ONNX is wrapping the exported module into a `ort.InferenceSession` but I can't find where the `ort.InferenceSession` is defined and how it could be passed down to the parent class and recognized as a `PreTrainedModel` that can be used in the pipeline later. Would you mind explaining a bit?

A bit context, for ExecuTorch, `ExecuTorchModule` is the ExecuTorch exported format that gets loaded from the local file system. The `.pte` file is sort of equivalent to the `.onnx` file from ONNX export.

Reply:
NVM, figured out a way to integrate and unblock